### PR TITLE
Allow income collection to send letters

### PR DESCRIPTION
**Pull request by richrace**
### Why

- Income Collection users need to send Letters
- Currently, only Leasehold users send letters

### What 

- Have two paths Leasehold/Income Collection for the different letter types

### Guidance for Review

- A lot of the code is duplicated on purpose and key fields changed `payment_ref` -> `tenancy_ref`
- JS has been rewritten to be less global
- More "Railsy" routes for the `IncomeCollection::LettersController`
- Removed the `preview_ajax` action from the `IncomeCollection::LettersController`

### Future Work 

- Minimise the code duplication. I haven't started doing any of that at the moment as I'm not 100% how the users will use this.

### CODE REVIEW

File: app/views/income_collection/letters/new.erb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        <span class="form-hint">Enter comma separated tenancy references</span>
+        <textarea class="form-control form-control-3-3" id="tenancy_refs" name="tenancy_refs" rows="6"></textarea>
+      </div>
+
+      <div class="form-group">
+        <label class="form-label" for="sms-template-select"><strong>Select template</strong></label>
+        <span class="form-hint">Select a letter template to send from the dropdown list below</span>
+        <%= select_tag(:template_id, options_for_select(@letter_templates
+                                                          .map { |template| [template.name, template.id] }
+                                                        ), id: 'sms-template-select', class:'form-control') %>
```
Comment (kjdchapman) [inline app/views/income_collection/letters/new.erb]: Does [`options_from_collection_for_select`](https://apidock.com/rails/ActionView/Helpers/FormOptionsHelper/options_from_collection_for_select) work here? Think it's a bit cleaner.

Can't do multi-line suggestions yet - something like 
```
select_tag(
  :template_id, 
  options_from_collection_for_select(@letter_templates, :id, :name)
)
```

File: app/views/income_collection/letters/preview.html.erb [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+         data-template-id="<%= @preview.dig(:template, :id) %>"
+         data-income-collection
+         data-url="<%= income_collection_letters_path %>"
+    >
+      <div class="grid-row">
+        <div class="column-full">
+          <h1>Letter preview</h1>
+          <strong>Template name: </strong><%= @preview.dig(:template,:name) %>
+          <div class="letter_preview">
+            <% if @preview[:document_id] %>
```
Comment (kjdchapman) [inline app/views/income_collection/letters/preview.html.erb]: Can we drop this conditional now we don't have to support two versions of the `income-api`?

File: app/views/income_collection/letters/preview.js.erb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+errors = '<%= @preview.dig(:errors).to_json %>';
+errors = JSON.parse(errors.replace(/&quot;/g, '"'));
+
+if (errors.length === 0) {
```
Comment (kjdchapman) [inline app/views/income_collection/letters/preview.js.erb]: :tada: very happy to see the scary "ERRORS: 0" red text go away when there aren't errors!

File: config/routes.rb [hunk 4]
```diff
@@ -5,11 +5,18 @@
 
   get '/search', to: 'search_tenancies#show'
 
-  get '/letters/new', to: 'letters#new'
-  get '/letters/preview', to: redirect('/letters/new')
-  post '/letters/preview', to: 'letters#preview', as: :letter_preview
-  post '/letters', to: 'letters#send_letter', as: :send_letter
-  post '/letters/ajax_preview', to: 'letters#ajax_preview', as: :ajax_preview
+  namespace :leasehold do
```
Comment (kjdchapman) [inline config/routes.rb]: :+1: I like the namespacing/folder structure changes around this split.

File: lib/hackney/income/get_letter_preview.rb [hunk 5]
```diff
@@ -5,9 +5,12 @@ def initialize(letters_gateway:)
         @letters_gateway = letters_gateway
       end
 
-      def execute(pay_ref:, template_id:, user:)
+      def execute(template_id:, user:, pay_ref: nil, tenancy_ref: nil)
+        raise ArgumentError, 'pay_ref or tenancy_ref must be supplied' if [pay_ref, tenancy_ref].all?(&:blank?)
```
Comment (kjdchapman) [inline lib/hackney/income/get_letter_preview.rb]: How does this differ from having no defaults on the named params? I think I'm missing something.

File: spec/controllers/income_collection/letters_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,162 @@
+require 'rails_helper'
+
+describe IncomeCollection::LettersController do
```
Comment (kjdchapman) [inline spec/controllers/income_collection/letters_controller_spec.rb]: I think this is copied across from the generic letters controller spec? If so, I'm fine with keeping it for now.

File: lib/hackney/income/get_letter_preview.rb (hunk 5 above)
Comment (kjdchapman) [reply lib/hackney/income/get_letter_preview.rb]: Ah, I think I get it - if the caller explicitly passes `nil`, this implementation still throws the error?

File: spec/controllers/income_collection/letters_controller_spec.rb (hunk 6 above)
Comment (kjdchapman) [reply spec/controllers/income_collection/letters_controller_spec.rb]: I see a bunch of acceptance specs around this behaviour as well, nice.

File: app/assets/javascripts/income_collection/preview.js [hunk 7]
```diff
@@ -0,0 +1,128 @@
+var IncomeCollectionPreview = (function() {
+  function IncomeCollectionPreview(element) {
+    var _this = this;
```
[the remark this answers is not in the document]
Comment (richrace) [reply app/assets/javascripts/income_collection/preview.js]: JS Contexts https://medium.com/@denny.headrick/this-bind-this-a-javascript-story-313ff84a12c8

I do not always want to `.bind(this)` to functions and want to always access the instance (`_this`) at any level.

File: app/assets/javascripts/income_collection/preview.js [hunk 8]
```diff
@@ -0,0 +1,128 @@
+var IncomeCollectionPreview = (function() {
```
[the remark this answers is not in the document]
Comment (richrace) [reply app/assets/javascripts/income_collection/preview.js]: https://hackney.atlassian.net/browse/MA-226

File: app/assets/javascripts/preview.js [hunk 9]
```diff
@@ -1,5 +1,5 @@
 $( document ).on('turbolinks:load', function() {
-  if ($('.letters').length > 0 ){
+  if ($('.letters[data-leasehold]').length > 0 ){
```
[the remark this answers is not in the document]
Comment (richrace) [reply app/assets/javascripts/preview.js]: https://hackney.atlassian.net/browse/MA-226

File: app/assets/stylesheets/components/main.scss [hunk 10]
```diff
@@ -75,6 +75,10 @@ main {
     &--right {
       float: right;
     }
+
+    &--mt-10 {
```
[the remark this answers is not in the document]
Comment (richrace) [reply app/assets/stylesheets/components/main.scss]: I would if I could but we not seem to be using the linked to project.

File: app/controllers/leasehold/letters_controller.rb [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+          format.html { redirect_to leasehold_letters_new_path, notice: 'Successfully sent' }
+          format.js   {}
+        end
+      end
+    end
+
+    private
+
+    def payment_refs_not_found?
+      !@preview[:preview].present? || @payment_refs.empty? || @preview[:status_code] == 404
```
[the remark this answers is not in the document]
Comment (richrace) [reply app/controllers/leasehold/letters_controller.rb]: I have tried to not purposely edit/change existing code unless required for functionality.

File: app/views/income_collection/letters/new.erb (hunk 1 above)
Comment (richrace) [reply app/views/income_collection/letters/new.erb]: I can change it if you'd like. I don't really have an opinion.

File: app/views/income_collection/letters/preview.html.erb (hunk 2 above)
Comment (richrace) [reply app/views/income_collection/letters/preview.html.erb]: If the letter errors, a `document` does not get created.

File: lib/hackney/income/get_letter_preview.rb (hunk 5 above)
Comment (richrace) [reply lib/hackney/income/get_letter_preview.rb]: Correct. One *must* be supplied.

File: spec/controllers/income_collection/letters_controller_spec.rb (hunk 6 above)
Comment (richrace) [reply spec/controllers/income_collection/letters_controller_spec.rb]: > I think this is copied across from the generic letters controller spec? If so, I'm fine with keeping it for now.

Yup.

**Pull request merged**